README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

# Onyxsio Grid View

Provides a Flutter grids layout.

## Getting started

In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:
  onyxsio_grid_view: <latest_version>
```

In your library add the following import:

```dart
import 'package:onyxsio_grid_view/onyxsio_grid_view.dart';
```

For help getting started with Flutter, view the online [documentation][flutter_documentation].

## Features

This layout facilitates the browsing of uncropped peer content. Container heights are sized based on the widget size.

### UI

![Staired Grid Layout][staired_preview]

## Usage

Below you'll find the code to create this grid layout:

```dart
      OnyxsioGridView.builder(
        padding: const EdgeInsets.all(8),
        itemCount: 20,
        physics: const BouncingScrollPhysics(),
        staggeredTileBuilder: (index) => const OnyxsioStaggeredTile.fit(2),
        crossAxisCount: 4,
        mainAxisSpacing: 10,
        crossAxisSpacing: 10,
        shrinkWrap: true,
        itemBuilder: (context, index) => OnyxsioGridTile(
          index: index,
          heightList: const [300, 220, 220, 520],
          child: Container(color: Colors.blue),
        ),
      )
```

### Optional information

```dart
staggeredTileBuilder: (index) =>
  OnyxsioStaggeredTile.count(2, index.isEven ? 2 : 1),

```

### Optional information

```dart
staggeredTileBuilder: (index) =>
  OnyxsioStaggeredTile.extent(2, index.isEven ? 200 : 50),
```

## Additional information

I'm working on my packages on my free-time, but I don't have as much time as I would. If this package or any other package I created is helping you, please consider to sponsor me so that I can take time to read the issues, fix bugs, merge pull requests and add features to these packages.

<!-- [![Pub][pub_badge]][pub] [![BuyMeACoffee][buy_me_a_coffee_badge]][buy_me_a_coffee] -->

## Contributions

Feel free to contribute to this project.

If you find a bug or want a feature, but don't know how to fix/implement it, please fill an [issue][issue].  
If you fixed a bug or implemented a feature, please send a [pull request][pr].

<!-- Links -->

[issue]: https://github.com/sudeshnb/onyxsio_grid_view/issues
[pr]: https://github.com/sudeshnb/onyxsio_grid_view/pulls
[flutter_documentation]: https://docs.flutter.dev/
[pub]: https://pub.dartlang.org/packages/onyxsio_grid_view
[staired_preview]: https://user-images.githubusercontent.com/33403844/214500635-7860d799-67c7-4f98-b031-725bc0e1922c.png

<!-- [buy_me_a_coffee]: https://www.buymeacoffee.com/sudeshnb -->
<!-- [buy_me_a_coffee_badge]: https://user-images.githubusercontent.com/33403844/214502169-982df8a4-a758-44e7-8c0f-cc85fd29547e.svg -->


CHANGELOG.md
# 1.0.4

- **Breaking change:** Added an example for the Package.
- Fixed `example/main.dart`.
- Fixed `doc/opening.md`.

# 1.0.3

<!-- - **Breaking change:** Removed deprecated `sayHello()` method. -->

- Remove The unnecessary package because all of the used elements are also provided by the other package.
- Fixed lib/src/onyxsio.dart:3:8.
- Initial stable release.

<!-- ## Upgrading from 0.1.x -->

<!-- Change all calls to `sayHello()` to instead be to `sayHi()`. -->

<!-- # 1.0.1 -->

<!-- - Deprecated the `sayHello()` method; use `sayHi()` instead. -->

# 1.0.0

- Initial development release.


test/onyxsio_grid_view_test.dart
// import 'package:flutter/widgets.dart';

// import 'package:onyxsio_grid_view/onyxsio_grid_view.dart';

// Size getTileSize(OnyxsioStaggeredTile tile, double cellLength) {
//   return Size(
//     tile.crossAxisCellCount * cellLength,
//     tile.mainAxisExtent ?? tile.mainAxisCellCount! * cellLength,
//   );
// }

void main() {}


lib/src/onyxsio.dart
import 'dart:collection';

// import 'package:flutter/foundation.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter/widgets.dart';
import 'onyxsio_staggered_grid.dart';
import 'onyxsio_variable_size_box_adaptor.dart';

/// Helps subclasses build their children lazily using a [SliverVariableSizeChildDelegate].
abstract class OnyxsioVariableSizeBoxAdaptorWidget
    extends SliverWithKeepAliveWidget {
  /// Initializes fields for subclasses.
  const OnyxsioVariableSizeBoxAdaptorWidget({
    Key? key,
    required this.delegate,
  }) : super(key: key);

  final SliverChildDelegate delegate;

  @override
  OnyxsioVariableSizeBoxAdaptorElement createElement() =>
      OnyxsioVariableSizeBoxAdaptorElement(this);

  @override
  RenderOnyxsioVariableSizeBoxAdaptor createRenderObject(BuildContext context);

  double? estimateMaxScrollOffset(
    SliverConstraints constraints,
    int firstIndex,
    int lastIndex,
    double leadingScrollOffset,
    double trailingScrollOffset,
  ) {
    assert(lastIndex >= firstIndex);
    return delegate.estimateMaxScrollOffset(
      firstIndex,
      lastIndex,
      leadingScrollOffset,
      trailingScrollOffset,
    );
  }

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(
      DiagnosticsProperty<SliverChildDelegate>('delegate', delegate),
    );
  }
}

class OnyxsioVariableSizeBoxAdaptorElement extends RenderObjectElement
    implements RenderOnyxsioVariableSizeBoxChildManager {
  /// Creates an element that lazily builds children for the given widget.
  OnyxsioVariableSizeBoxAdaptorElement(
      OnyxsioVariableSizeBoxAdaptorWidget widget)
      : super(widget);

  @override
  OnyxsioVariableSizeBoxAdaptorWidget get widget =>
      super.widget as OnyxsioVariableSizeBoxAdaptorWidget;

  @override
  RenderOnyxsioVariableSizeBoxAdaptor get renderObject =>
      super.renderObject as RenderOnyxsioVariableSizeBoxAdaptor;

  @override
  void update(covariant OnyxsioVariableSizeBoxAdaptorWidget newWidget) {
    final OnyxsioVariableSizeBoxAdaptorWidget oldWidget = widget;
    super.update(newWidget);
    final SliverChildDelegate newDelegate = newWidget.delegate;
    final SliverChildDelegate oldDelegate = oldWidget.delegate;
    if (newDelegate != oldDelegate &&
        (newDelegate.runtimeType != oldDelegate.runtimeType ||
            newDelegate.shouldRebuild(oldDelegate))) {
      performRebuild();
    }
  }

  final Map<int, Widget?> _childWidgets = HashMap<int, Widget?>();
  final SplayTreeMap<int, Element> _childElements =
      SplayTreeMap<int, Element>();

  @override
  void performRebuild() {
    _childWidgets.clear(); // Reset the cache, as described above.
    super.performRebuild();
    assert(_currentlyUpdatingChildIndex == null);
    try {
      late final int firstIndex;
      late final int lastIndex;
      if (_childElements.isEmpty) {
        firstIndex = 0;
        lastIndex = 0;
      } else if (_didUnderflow) {
        firstIndex = _childElements.firstKey()!;
        lastIndex = _childElements.lastKey()! + 1;
      } else {
        firstIndex = _childElements.firstKey()!;
        lastIndex = _childElements.lastKey()!;
      }

      for (int index = firstIndex; index <= lastIndex; ++index) {
        _currentlyUpdatingChildIndex = index;
        final Element? newChild =
            updateChild(_childElements[index], _build(index), index);
        if (newChild != null) {
          _childElements[index] = newChild;
        } else {
          _childElements.remove(index);
        }
      }
    } finally {
      _currentlyUpdatingChildIndex = null;
    }
  }

  Widget? _build(int index) {
    return _childWidgets.putIfAbsent(
        index, () => widget.delegate.build(this, index));
  }

  @override
  void createChild(int index) {
    assert(_currentlyUpdatingChildIndex == null);
    owner!.buildScope(this, () {
      Element? newChild;
      try {
        _currentlyUpdatingChildIndex = index;
        newChild = updateChild(_childElements[index], _build(index), index);
      } finally {
        _currentlyUpdatingChildIndex = null;
      }
      if (newChild != null) {
        _childElements[index] = newChild;
      } else {
        _childElements.remove(index);
      }
    });
  }

  @override
  Element? updateChild(Element? child, Widget? newWidget, dynamic newSlot) {
    final oldParentData = child?.renderObject?.parentData
        as OnyxsioVariableSizeBoxAdaptorParentData?;
    final Element? newChild = super.updateChild(child, newWidget, newSlot);
    final newParentData = newChild?.renderObject?.parentData
        as OnyxsioVariableSizeBoxAdaptorParentData?;

    // set keepAlive to true in order to populate the cache
    if (newParentData != null) {
      newParentData.keepAlive = true;
    }

    // Preserve the old layoutOffset if the renderObject was swapped out.
    if (oldParentData != newParentData &&
        oldParentData != null &&
        newParentData != null) {
      newParentData.layoutOffset = oldParentData.layoutOffset;
    }

    return newChild;
  }

  @override
  void forgetChild(Element child) {
    assert(child.slot != null);
    assert(_childElements.containsKey(child.slot));
    _childElements.remove(child.slot);
    super.forgetChild(child);
  }

  @override
  void removeChild(RenderBox child) {
    final int index = renderObject.indexOf(child);
    assert(_currentlyUpdatingChildIndex == null);
    assert(index >= 0);
    owner!.buildScope(this, () {
      assert(_childElements.containsKey(index));
      try {
        _currentlyUpdatingChildIndex = index;
        final Element? result = updateChild(_childElements[index], null, index);
        assert(result == null);
      } finally {
        _currentlyUpdatingChildIndex = null;
      }
      _childElements.remove(index);
      assert(!_childElements.containsKey(index));
    });
  }

  double? _extrapolateMaxScrollOffset(
    int? firstIndex,
    int? lastIndex,
    double? leadingScrollOffset,
    double? trailingScrollOffset,
  ) {
    final int? childCount = widget.delegate.estimatedChildCount;
    if (childCount == null) {
      return double.infinity;
    }
    if (lastIndex == childCount - 1) {
      return trailingScrollOffset;
    }
    final int reifiedCount = lastIndex! - firstIndex! + 1;
    final double averageExtent =
        (trailingScrollOffset! - leadingScrollOffset!) / reifiedCount;
    final int remainingCount = childCount - lastIndex - 1;
    return trailingScrollOffset + averageExtent * remainingCount;
  }

  @override
  double estimateMaxScrollOffset(
    SliverConstraints constraints, {
    int? firstIndex,
    int? lastIndex,
    double? leadingScrollOffset,
    double? trailingScrollOffset,
  }) {
    return widget.estimateMaxScrollOffset(
          constraints,
          firstIndex!,
          lastIndex!,
          leadingScrollOffset!,
          trailingScrollOffset!,
        ) ??
        _extrapolateMaxScrollOffset(
          firstIndex,
          lastIndex,
          leadingScrollOffset,
          trailingScrollOffset,
        )!;
  }

  @override
  int get childCount => widget.delegate.estimatedChildCount ?? 0;

  @override
  void didStartLayout() {
    assert(debugAssertChildListLocked());
  }

  @override
  void didFinishLayout() {
    assert(debugAssertChildListLocked());
    final int firstIndex = _childElements.firstKey() ?? 0;
    final int lastIndex = _childElements.lastKey() ?? 0;
    widget.delegate.didFinishLayout(firstIndex, lastIndex);
  }

  int? _currentlyUpdatingChildIndex;

  @override
  bool debugAssertChildListLocked() {
    assert(_currentlyUpdatingChildIndex == null);
    return true;
  }

  @override
  void didAdoptChild(RenderBox child) {
    assert(_currentlyUpdatingChildIndex != null);
    final childParentData =
        child.parentData! as OnyxsioVariableSizeBoxAdaptorParentData;
    childParentData.index = _currentlyUpdatingChildIndex;
  }

  bool _didUnderflow = false;

  @override
  void setDidUnderflow(bool value) {
    _didUnderflow = value;
  }

  @override
  void insertRenderObjectChild(covariant RenderBox child, int slot) {
    assert(_currentlyUpdatingChildIndex == slot);
    assert(renderObject.debugValidateChild(child));
    renderObject[_currentlyUpdatingChildIndex!] = child;
    assert(() {
      final childParentData =
          child.parentData! as OnyxsioVariableSizeBoxAdaptorParentData;
      assert(slot == childParentData.index);
      return true;
    }());
  }

  @override
  void moveRenderObjectChild(
    covariant RenderObject child,
    covariant Object? oldSlot,
    covariant Object? newSlot,
  ) {
    assert(false);
  }

  @override
  void removeRenderObjectChild(
    covariant RenderObject child,
    covariant Object? slot,
  ) {
    assert(_currentlyUpdatingChildIndex != null);
    renderObject.remove(_currentlyUpdatingChildIndex!);
  }

  @override
  void visitChildren(ElementVisitor visitor) {
    // The toList() is to make a copy so that the underlying list can be modified by
    // the visitor:
    _childElements.values.toList().forEach(visitor);
  }

  @override
  void debugVisitOnstageChildren(ElementVisitor visitor) {
    _childElements.values.where((Element child) {
      final parentData =
          child.renderObject!.parentData as SliverMultiBoxAdaptorParentData?;
      late double itemExtent;
      switch (renderObject.constraints.axis) {
        case Axis.horizontal:
          itemExtent = child.renderObject!.paintBounds.width;
          break;
        case Axis.vertical:
          itemExtent = child.renderObject!.paintBounds.height;
          break;
      }

      return parentData!.layoutOffset! <
              renderObject.constraints.scrollOffset +
                  renderObject.constraints.remainingPaintExtent &&
          parentData.layoutOffset! + itemExtent >
              renderObject.constraints.scrollOffset;
    }).forEach(visitor);
  }
}

class OnyxsioStaggeredGrid extends OnyxsioVariableSizeBoxAdaptorWidget {
  const OnyxsioStaggeredGrid({
    Key? key,
    required SliverChildDelegate delegate,
    required this.gridDelegate,
  }) : super(key: key, delegate: delegate);

  // OnyxsioStaggeredGrid.count({
  //   Key? key,
  //   required int crossAxisCount,
  //   double mainAxisSpacing = 0.0,
  //   double crossAxisSpacing = 0.0,
  //   List<Widget> children = const <Widget>[],
  //   List<StaggeredTile> staggeredTiles = const <StaggeredTile>[],
  // })  : gridDelegate = SliverStaggeredGridDelegateWithFixedCrossAxisCount(
  //         crossAxisCount: crossAxisCount,
  //         mainAxisSpacing: mainAxisSpacing,
  //         crossAxisSpacing: crossAxisSpacing,
  //         staggeredTileBuilder: (i) => staggeredTiles[i],
  //         staggeredTileCount: staggeredTiles.length,
  //       ),
  //       super(
  //         key: key,
  //         delegate: SliverChildListDelegate(
  //           children,
  //         ),
  //       );

  OnyxsioStaggeredGrid.builder({
    Key? key,
    required int crossAxisCount,
    required OnyxsioIndexedStaggeredTileBuilder staggeredTileBuilder,
    required IndexedWidgetBuilder itemBuilder,
    required int itemCount,
    double mainAxisSpacing = 0,
    double crossAxisSpacing = 0,
  })  : gridDelegate = SliverStaggeredGridDelegateWithFixedCrossAxisCount(
          crossAxisCount: crossAxisCount,
          mainAxisSpacing: mainAxisSpacing,
          crossAxisSpacing: crossAxisSpacing,
          staggeredTileBuilder: staggeredTileBuilder,
          staggeredTileCount: itemCount,
        ),
        super(
          key: key,
          delegate: SliverChildBuilderDelegate(
            itemBuilder,
            childCount: itemCount,
          ),
        );

  // OnyxsioStaggeredGrid.extent({
  //   Key? key,
  //   required double maxCrossAxisExtent,
  //   double mainAxisSpacing = 0,
  //   double crossAxisSpacing = 0,
  //   List<Widget> children = const <Widget>[],
  //   List<StaggeredTile> staggeredTiles = const <StaggeredTile>[],
  // })  : gridDelegate = SliverStaggeredGridDelegateWithMaxCrossAxisExtent(
  //         maxCrossAxisExtent: maxCrossAxisExtent,
  //         mainAxisSpacing: mainAxisSpacing,
  //         crossAxisSpacing: crossAxisSpacing,
  //         staggeredTileBuilder: (i) => staggeredTiles[i],
  //         staggeredTileCount: staggeredTiles.length,
  //       ),
  //       super(
  //         key: key,
  //         delegate: SliverChildListDelegate(
  //           children,
  //         ),
  //       );

  // OnyxsioStaggeredGrid.extentBuilder({
  //   Key? key,
  //   required double maxCrossAxisExtent,
  //   required IndexedStaggeredTileBuilder staggeredTileBuilder,
  //   required IndexedWidgetBuilder itemBuilder,
  //   required int itemCount,
  //   double mainAxisSpacing = 0,
  //   double crossAxisSpacing = 0,
  // })  : gridDelegate = SliverStaggeredGridDelegateWithMaxCrossAxisExtent(
  //         maxCrossAxisExtent: maxCrossAxisExtent,
  //         mainAxisSpacing: mainAxisSpacing,
  //         crossAxisSpacing: crossAxisSpacing,
  //         staggeredTileBuilder: staggeredTileBuilder,
  //         staggeredTileCount: itemCount,
  //       ),
  //       super(
  //         key: key,
  //         delegate: SliverChildBuilderDelegate(
  //           itemBuilder,
  //           childCount: itemCount,
  //         ),
  //       );

  /// The delegate that controls the size and position of the children.
  final OnyxsioStaggeredGridDelegate gridDelegate;

  @override
  RenderOnyxsioStaggeredGrid createRenderObject(BuildContext context) {
    final element = context as OnyxsioVariableSizeBoxAdaptorElement;
    return RenderOnyxsioStaggeredGrid(
        childManager: element, gridDelegate: gridDelegate);
  }

  @override
  void updateRenderObject(
      BuildContext context, RenderOnyxsioStaggeredGrid renderObject) {
    renderObject.gridDelegate = gridDelegate;
  }
}


lib/onyxsio_grid_view.dart
// ignore_for_file: must_be_immutable

library onyxsio_grid_view;

import 'package:flutter/widgets.dart';
import 'src/onyxsio.dart';
import 'src/onyxsio_staggered_grid.dart';
export 'src/staggered_tile.dart';

class OnyxsioGridView extends BoxScrollView {
  OnyxsioGridView({
    Key? key,
    Axis scrollDirection = Axis.vertical,
    bool reverse = false,
    ScrollController? controller,
    bool? primary,
    ScrollPhysics? physics,
    bool shrinkWrap = false,
    EdgeInsetsGeometry? padding,
    required this.gridDelegate,
    bool addAutomaticKeepAlives = true,
    bool addRepaintBoundaries = true,
    List<Widget> children = const <Widget>[],
    String? restorationId,
  })  : childrenDelegate = SliverChildListDelegate(
          children,
          addAutomaticKeepAlives: addAutomaticKeepAlives,
          addRepaintBoundaries: addRepaintBoundaries,
        ),
        super(
          key: key,
          scrollDirection: scrollDirection,
          reverse: reverse,
          controller: controller,
          primary: primary,
          physics: physics,
          shrinkWrap: shrinkWrap,
          padding: padding,
          restorationId: restorationId,
        );

  OnyxsioGridView.builder({
    Key? key,
    Axis scrollDirection = Axis.vertical,
    bool reverse = false,
    ScrollController? controller,
    bool? primary,
    ScrollPhysics? physics,
    bool shrinkWrap = false,
    EdgeInsetsGeometry? padding,
    required int crossAxisCount,
    required OnyxsioIndexedWidgetBuilder itemBuilder,
    required OnyxsioIndexedStaggeredTileBuilder staggeredTileBuilder,
    int? itemCount,
    double mainAxisSpacing = 0.0,
    double crossAxisSpacing = 0.0,
    bool addAutomaticKeepAlives = true,
    bool addRepaintBoundaries = true,
    String? restorationId,
  })  : gridDelegate = SliverStaggeredGridDelegateWithFixedCrossAxisCount(
          crossAxisCount: crossAxisCount,
          mainAxisSpacing: mainAxisSpacing,
          crossAxisSpacing: crossAxisSpacing,
          staggeredTileBuilder: staggeredTileBuilder,
          staggeredTileCount: itemCount,
        ),
        childrenDelegate = SliverChildBuilderDelegate(
          itemBuilder,
          childCount: itemCount,
          addAutomaticKeepAlives: addAutomaticKeepAlives,
          addRepaintBoundaries: addRepaintBoundaries,
        ),
        super(
          key: key,
          scrollDirection: scrollDirection,
          reverse: reverse,
          controller: controller,
          primary: primary,
          physics: physics,
          shrinkWrap: shrinkWrap,
          padding: padding,
          restorationId: restorationId,
        );

  final OnyxsioStaggeredGridDelegate gridDelegate;

  final SliverChildDelegate childrenDelegate;

  @override
  Widget buildChildLayout(BuildContext context) {
    return OnyxsioStaggeredGrid(
      delegate: childrenDelegate,
      gridDelegate: gridDelegate,
    );
  }
}

// typedef IndexedWidgetBuilder = Widget Function(BuildContext context, int index);

typedef OnyxsioIndexedWidgetBuilder = OnyxsioGridTile Function(
    BuildContext context, int index);

class OnyxsioGridTile extends StatelessWidget {
  OnyxsioGridTile(
      {Key? key,
      this.defaultHeight = 200.0,
      required this.heightList,
      required this.child,
      required this.index})
      : super(key: key);
  int index;
  Widget child;
  double defaultHeight;
  List<double> heightList;

  @override
  Widget build(BuildContext context) {
    return Container(
      constraints: BoxConstraints(minHeight: getMinHeight(index)),
      child: child,
    );
  }

  /// To return different height for different widgets
  double getMinHeight(int index) {
    for (var i = 0; i < heightList.length; i++) {
      // if (index == (index % deived))
      if (index == i) {
        return heightList[i];
      }
    }
    return defaultHeight;
  }
}


lib/src/staggered_tile.dart
class OnyxsioStaggeredTile {
  const OnyxsioStaggeredTile.count(
      this.crossAxisCellCount, this.mainAxisCellCount)
      : assert(crossAxisCellCount >= 0),
        assert(mainAxisCellCount != null && mainAxisCellCount >= 0),
        mainAxisExtent = null;

  const OnyxsioStaggeredTile.extent(
      this.crossAxisCellCount, this.mainAxisExtent)
      : assert(crossAxisCellCount >= 0),
        assert(mainAxisExtent != null && mainAxisExtent >= 0),
        mainAxisCellCount = null;

  const OnyxsioStaggeredTile.fit(this.crossAxisCellCount)
      : assert(crossAxisCellCount >= 0),
        mainAxisExtent = null,
        mainAxisCellCount = null;

  /// The number of cells occupied in the cross axis.
  final int crossAxisCellCount;

  /// The number of cells occupied in the main axis.
  final double? mainAxisCellCount;

  /// The number of pixels occupied in the main axis.
  final double? mainAxisExtent;

  bool get fitContent => mainAxisCellCount == null && mainAxisExtent == null;
}


lib/src/tile_container_render_object_mixin.dart
import 'dart:collection';
import 'package:flutter/rendering.dart';

mixin TileContainerRenderObjectMixin<ChildType extends RenderObject,
    ParentDataType extends ParentData> on RenderObject {
  final SplayTreeMap<int, ChildType> _childRenderObjects =
      SplayTreeMap<int, ChildType>();

  /// The number of children.
  int get childCount => _childRenderObjects.length;

  Iterable<ChildType> get children => _childRenderObjects.values;

  Iterable<int> get indices => _childRenderObjects.keys;

  /// Always returns true.
  bool debugValidateChild(RenderObject child) {
    assert(() {
      if (child is! ChildType) {
        throw FlutterError(
            'A $runtimeType expected a child of type $ChildType but received a '
            'child of type ${child.runtimeType}.\n'
            'RenderObjects expect specific types of children because they '
            'coordinate with their children during layout and paint. For '
            'example, a RenderSliver cannot be the child of a RenderBox because '
            'a RenderSliver does not understand the RenderBox layout protocol.\n'
            '\n'
            'The $runtimeType that expected a $ChildType child was created by:\n'
            '  $debugCreator\n'
            '\n'
            'The ${child.runtimeType} that did not match the expected child type '
            'was created by:\n'
            '  ${child.debugCreator}\n');
      }
      return true;
    }());
    return true;
  }

  ChildType? operator [](int index) => _childRenderObjects[index];

  void operator []=(int index, ChildType child) {
    if (index < 0) {
      throw ArgumentError(index);
    }
    _removeChild(_childRenderObjects[index]);
    adoptChild(child);
    _childRenderObjects[index] = child;
  }

  void forEachChild(void Function(ChildType child) f) {
    _childRenderObjects.values.forEach(f);
  }

  /// Remove the child at the specified index from the child list.
  void remove(int index) {
    final child = _childRenderObjects.remove(index);
    _removeChild(child);
  }

  void _removeChild(ChildType? child) {
    if (child != null) {
      // Remove the old child.
      dropChild(child);
    }
  }

  /// Remove all their children from this render object's child list.
  ///
  /// More efficient than removing them individually.
  void removeAll() {
    _childRenderObjects.values.forEach(dropChild);
    _childRenderObjects.clear();
  }

  @override
  void attach(PipelineOwner owner) {
    super.attach(owner);
    for (var child in _childRenderObjects.values) {
      child.attach(owner);
    }
  }

  @override
  void detach() {
    super.detach();
    for (var child in _childRenderObjects.values) {
      child.detach();
    }
  }

  @override
  void redepthChildren() {
    _childRenderObjects.values.forEach(redepthChild);
  }

  @override
  void visitChildren(RenderObjectVisitor visitor) {
    _childRenderObjects.values.forEach(visitor);
  }

  @override
  List<DiagnosticsNode> debugDescribeChildren() {
    final List<DiagnosticsNode> children = <DiagnosticsNode>[];
    _childRenderObjects.forEach((index, child) =>
        children.add(child.toDiagnosticsNode(name: 'child $index')));
    return children;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
// import 'package:flutter/rendering.dart';
import 'package:onyxsio_grid_view/onyxsio_grid_view.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Onyxsio GridView builder',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Onyxsio GridView builder"),
      ),
      body: OnyxsioGridView.builder(
        padding: const EdgeInsets.all(8),
        itemCount: 20,
        physics: const BouncingScrollPhysics(),
        staggeredTileBuilder: (index) =>
            // OnyxsioStaggeredTile.count(2, index.isEven ? 2 : 1),
            // OnyxsioStaggeredTile.extent(2, index.isEven ? 200 : 50),
            const OnyxsioStaggeredTile.fit(2),
        crossAxisCount: 4,
        mainAxisSpacing: 10,
        crossAxisSpacing: 10,
        shrinkWrap: true,
        itemBuilder: (context, index) => OnyxsioGridTile(
          index: index,
          heightList: const [300, 220, 220, 520],
          child: Container(color: Colors.blue),
        ),
      ),
    );
  }
}
